redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+
+#[derive(PartialEq, Debug, Clone)]
+pub(crate) struct ScanState {
+    // the real cursor in the scan operation
+    cursor: u64,
+    // a map of the slots that have been scanned
+    scanned_slots_map: SlotsBitsArray,
+    // the address that is being scanned currently, based on the next slot set to 0 in the scanned_slots_map, and the address "own" the slot
+    // in the SlotMap
+    pub(crate) address_in_scan: String,
+    // epoch represent the version of the address, when a failover happens or slots migrate in the epoch will be updated to +1
+    address_epoch: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ClusterScanArgs {
+    pub(crate) scan_state_cursor: ScanStateCursor,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<ObjectType>,
+}
+
+#[derive(Debug, Clone)]
+/// Represents the type of an object in Redis.
+pub enum ObjectType {
+    /// Represents a string object in Redis.
+    STRING,
+    /// Represents a list object in Redis.
+    LIST,
+    /// Represents a set object in Redis.
+    SET,
+    /// Represents a sorted set object in Redis.
+    ZSET,
+    /// Represents a hash object in Redis.
+    HASH,
+}
+
+impl std::fmt::Display for ObjectType {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            ObjectType::STRING => write!(f, "string"),
+            ObjectType::LIST => write!(f, "list"),
+            ObjectType::SET => write!(f, "set"),
+            ObjectType::ZSET => write!(f, "zset"),
+            ObjectType::HASH => write!(f, "hash"),
+        }
+    }
+}
+
+impl ClusterScanArgs {
+    pub(crate) fn new(
+        scan_state_cursor: ScanStateCursor,
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<ObjectType>,
+    ) -> Self {
+        Self {
+            scan_state_cursor,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateCursor {
+    scan_state_ref: Arc<Option<ScanState>>,
+}
+
+impl ScanStateCursor {
+    /// Creates a new instance of `ScanStateCursor` from a given `ScanState`.
+    fn from_scan_state(scan_state: &ScanState) -> Self {
+        Self {
+            scan_state_ref: Arc::new(Some(scan_state.clone())),
+        }
+    }
+
+    /// Creates a new instance of `ScanStateCursor`.
+    ///
+    /// This method initializes the `ScanStateCursor` with a reference to a `ScanState` that is initially set to `None`.
+    /// An empty ScanState is equivilant to a 0 cursor.
+    pub fn new() -> Self {
+        Self {
+            scan_state_ref: Arc::new(None),
+        }
+    }
+
+    /// Returns `true` if the scan state is empty, indicating that the scan operation has finished or has not started.
+    pub fn is_none(&self) -> bool {
```

Comment:
Nit: you could change the order: `Option<Arc<ScanState>>` and avoid re-implementing `is_none`, `is_some` etc

Reply:
Ill need to implement it anyway because the ref is inside the cursor, and the user dont have access to it, so i need to create pub function checking it for him